File: tdz/src/idea.rs
use crate::{models::*, error::*};
use std::collections::HashMap;
use chrono::Utc;
pub struct IdeaManager {
    pub ideas: HashMap<String, Idea>,
    pub idea_tags: HashMap<String, Vec<String>>,
}
impl IdeaManager {
    pub fn new() -> Self {
        Self {
            ideas: HashMap::new(),
            idea_tags: HashMap::new(),
        }
    }
    pub async fn create_idea(&mut self, mut idea: Idea) -> Result<String> {
        idea.id = uuid::Uuid::new_v4().to_string();
        idea.created_at = Utc::now();
        idea.updated_at = Utc::now();
        self.idea_tags.insert(idea.id.clone(), idea.tags.clone());
        self.ideas.insert(idea.id.clone(), idea.clone());
        Ok(idea.id)
    }
    pub fn get_idea(&self, idea_id: &str) -> Option<&Idea> {
        self.ideas.get(idea_id)
    }
    pub fn get_all_ideas(&self) -> Vec<&Idea> {
        self.ideas.values().collect()
    }
    pub async fn update_idea(
        &mut self,
        idea_id: &str,
        updates: IdeaUpdate,
    ) -> Result<()> {
        if let Some(idea) = self.ideas.get_mut(idea_id) {
            if let Some(idea_text) = updates.idea {
                idea.idea = idea_text;
            }
            if let Some(share) = updates.share {
                idea.share = share;
            }
            if let Some(importance) = updates.importance {
                idea.importance = importance;
            }
            if let Some(tags) = updates.tags {
                idea.tags = tags.clone();
                self.idea_tags.insert(idea_id.to_string(), tags);
            }
            if let Some(context) = updates.context {
                idea.context = Some(context);
            }
            idea.updated_at = Utc::now();
        } else {
            return Err(TodoziError::ValidationError {
                message: format!("Idea {} not found", idea_id),
            });
        }
        Ok(())
    }
    pub async fn delete_idea(&mut self, idea_id: &str) -> Result<()> {
        if self.ideas.remove(idea_id).is_some() {
            self.idea_tags.remove(idea_id);
            Ok(())
        } else {
            Err(TodoziError::ValidationError {
                message: format!("Idea {} not found", idea_id),
            })
        }
    }
    pub fn search_ideas(&self, query: &str) -> Vec<&Idea> {
        let query_lower = query.to_lowercase();
        self.ideas
            .values()
            .filter(|idea| {
                idea.idea.to_lowercase().contains(&query_lower)
                    || idea
                        .tags
                        .iter()
                        .any(|tag| tag.to_lowercase().contains(&query_lower))
                    || if let Some(context) = &idea.context {
                        context.to_lowercase().contains(&query_lower)
                    } else {
                        false
                    }
            })
            .collect()
    }
    pub fn get_ideas_by_importance(&self, importance: IdeaImportance) -> Vec<&Idea> {
        self.ideas.values().filter(|idea| idea.importance == importance).collect()
    }
    pub fn get_ideas_by_share_level(&self, share_level: ShareLevel) -> Vec<&Idea> {
        self.ideas.values().filter(|idea| idea.share == share_level).collect()
    }
    pub fn get_ideas_by_tag(&self, tag: &str) -> Vec<&Idea> {
        let tag_lower = tag.to_lowercase();
        self.ideas
            .values()
            .filter(|idea| idea.tags.iter().any(|t| t.to_lowercase() == tag_lower))
            .collect()
    }
    pub fn get_public_ideas(&self) -> Vec<&Idea> {
        self.get_ideas_by_share_level(ShareLevel::Public)
    }
    pub fn get_team_ideas(&self) -> Vec<&Idea> {
        self.get_ideas_by_share_level(ShareLevel::Team)
    }
    pub fn get_private_ideas(&self) -> Vec<&Idea> {
        self.get_ideas_by_share_level(ShareLevel::Private)
    }
    pub fn get_breakthrough_ideas(&self) -> Vec<&Idea> {
        self.get_ideas_by_importance(IdeaImportance::Breakthrough)
    }
    pub fn get_recent_ideas(&self, limit: usize) -> Vec<&Idea> {
        let mut ideas: Vec<&Idea> = self.ideas.values().collect();
        ideas.sort_by(|a, b| b.created_at.cmp(&a.created_at));
        ideas.into_iter().take(limit).collect()
    }
    pub fn get_all_tags(&self) -> Vec<String> {
        let mut all_tags = std::collections::HashSet::new();
        for tags in self.idea_tags.values() {
            for tag in tags {
                all_tags.insert(tag.clone());
            }
        }
        all_tags.into_iter().collect()
    }
    pub fn get_tag_statistics(&self) -> HashMap<String, usize> {
        let mut stats = HashMap::new();
        for tags in self.idea_tags.values() {
            for tag in tags {
                *stats.entry(tag.clone()).or_insert(0) += 1;
            }
        }
        stats
    }
    pub fn get_idea_statistics(&self) -> IdeaStatistics {
        let total_ideas = self.ideas.len();
        let public_ideas = self.get_public_ideas().len();
        let team_ideas = self.get_team_ideas().len();
        let private_ideas = self.get_private_ideas().len();
        let breakthrough_ideas = self.get_breakthrough_ideas().len();
        let unique_tags = self.get_all_tags().len();
        IdeaStatistics {
            total_ideas,
            public_ideas,
            team_ideas,
            private_ideas,
            breakthrough_ideas,
            unique_tags,
        }
    }
}
#[derive(Debug, Clone)]
pub struct IdeaUpdate {
    pub idea: Option<String>,
    pub share: Option<ShareLevel>,
    pub importance: Option<IdeaImportance>,
    pub tags: Option<Vec<String>>,
    pub context: Option<String>,
}
impl IdeaUpdate {
    pub fn new() -> Self {
        Self {
            idea: None,
            share: None,
            importance: None,
            tags: None,
            context: None,
        }
    }
    pub fn idea(mut self, idea: String) -> Self {
        self.idea = Some(idea);
        self
    }
    pub fn share(mut self, share: ShareLevel) -> Self {
        self.share = Some(share);
        self
    }
    pub fn importance(mut self, importance: IdeaImportance) -> Self {
        self.importance = Some(importance);
        self
    }
    pub fn tags(mut self, tags: Vec<String>) -> Self {
        self.tags = Some(tags);
        self
    }
    pub fn context(mut self, context: String) -> Self {
        self.context = Some(context);
        self
    }
}
#[derive(Debug, Clone)]
pub struct IdeaStatistics {
    pub total_ideas: usize,
    pub public_ideas: usize,
    pub team_ideas: usize,
    pub private_ideas: usize,
    pub breakthrough_ideas: usize,
    pub unique_tags: usize,
}
impl IdeaStatistics {
    pub fn public_percentage(&self) -> f64 {
        if self.total_ideas == 0 {
            0.0
        } else {
            (self.public_ideas as f64 / self.total_ideas as f64) * 100.0
        }
    }
    pub fn team_percentage(&self) -> f64 {
        if self.total_ideas == 0 {
            0.0
        } else {
            (self.team_ideas as f64 / self.total_ideas as f64) * 100.0
        }
    }
    pub fn private_percentage(&self) -> f64 {
        if self.total_ideas == 0 {
            0.0
        } else {
            (self.private_ideas as f64 / self.total_ideas as f64) * 100.0
        }
    }
    pub fn breakthrough_percentage(&self) -> f64 {
        if self.total_ideas == 0 {
            0.0
        } else {
            (self.breakthrough_ideas as f64 / self.total_ideas as f64) * 100.0
        }
    }
}
pub fn parse_idea_format(idea_text: &str) -> Result<Idea> {
    let start_tag = "<idea>";
    let end_tag = "</idea>";
    let start = idea_text
        .find(start_tag)
        .ok_or_else(|| TodoziError::ValidationError {
            message: "Missing <idea> start tag".to_string(),
        })?;
    let end = idea_text
        .find(end_tag)
        .ok_or_else(|| TodoziError::ValidationError {
            message: "Missing </idea> end tag".to_string(),
        })?;
    let content = &idea_text[start + start_tag.len()..end];
    let parts: Vec<&str> = content.split(';').map(|s| s.trim()).collect();
    if parts.len() < 3 {
        return Err(TodoziError::ValidationError {
            message: "Invalid idea format: need at least 3 parts (idea; share; importance)"
                .to_string(),
        });
    }
    let share = match parts[1].to_lowercase().as_str() {
        "share" => ShareLevel::Public,
        "dont share" | "don't share" | "private" => ShareLevel::Private,
        "team" => ShareLevel::Team,
        _ => ShareLevel::Private,
    };
    let tags = if parts.len() > 3 && !parts[3].is_empty() {
        parts[3].split(',').map(|s| s.trim().to_string()).collect()
    } else {
        Vec::new()
    };
    let context = if parts.len() > 4 && !parts[4].is_empty() {
        Some(parts[4].to_string())
    } else {
        None
    };
    Ok(Idea {
        id: uuid::Uuid::new_v4().to_string(),
        idea: parts[0].to_string(),
        project_id: None,
        status: ItemStatus::Active,
        share,
        importance: parts[2]
            .parse::<IdeaImportance>()
            .map_err(|_| TodoziError::ValidationError {
                message: "Invalid idea importance".to_string(),
            })?,
        tags,
        context,
        created_at: Utc::now(),
        updated_at: Utc::now(),
    })
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_idea_manager_creation() {
        let manager = IdeaManager::new();
        assert_eq!(manager.ideas.len(), 0);
        assert_eq!(manager.idea_tags.len(), 0);
    }
    #[test]
    fn test_idea_update_builder() {
        let update = IdeaUpdate::new()
            .idea("New idea".to_string())
            .share(ShareLevel::Public)
            .importance(IdeaImportance::High);
        assert_eq!(update.idea, Some("New idea".to_string()));
        assert_eq!(update.share, Some(ShareLevel::Public));
        assert_eq!(update.importance, Some(IdeaImportance::High));
    }
    #[test]
    fn test_idea_statistics_percentages() {
        let stats = IdeaStatistics {
            total_ideas: 10,
            public_ideas: 4,
            team_ideas: 3,
            private_ideas: 3,
            breakthrough_ideas: 2,
            unique_tags: 8,
        };
        assert_eq!(stats.public_percentage(), 40.0);
        assert_eq!(stats.team_percentage(), 30.0);
        assert_eq!(stats.private_percentage(), 30.0);
        assert_eq!(stats.breakthrough_percentage(), 20.0);
        let empty_stats = IdeaStatistics {
            total_ideas: 0,
            public_ideas: 0,
            team_ideas: 0,
            private_ideas: 0,
            breakthrough_ideas: 0,
            unique_tags: 0,
        };
        assert_eq!(empty_stats.public_percentage(), 0.0);
        assert_eq!(empty_stats.team_percentage(), 0.0);
        assert_eq!(empty_stats.private_percentage(), 0.0);
        assert_eq!(empty_stats.breakthrough_percentage(), 0.0);
    }
    #[test]
    fn test_parse_idea_format() {
        let idea_text = "<idea>Use microservices for better scalability; share; high; architecture,microservices,scalability; This will improve deployment speed</idea>";
        let idea = parse_idea_format(idea_text).unwrap();
        assert_eq!(idea.idea, "Use microservices for better scalability");
        assert_eq!(idea.share, ShareLevel::Public);
        assert_eq!(idea.importance, IdeaImportance::High);
        assert_eq!(idea.tags, vec!["architecture", "microservices", "scalability"]);
        assert_eq!(idea.context, Some("This will improve deployment speed".to_string()));
    }
    #[test]
    fn test_parse_idea_format_minimal() {
        let idea_text = "<idea>Simple idea; private; low</idea>";
        let idea = parse_idea_format(idea_text).unwrap();
        assert_eq!(idea.idea, "Simple idea");
        assert_eq!(idea.share, ShareLevel::Private);
        assert_eq!(idea.importance, IdeaImportance::Low);
        assert_eq!(idea.tags.len(), 0);
        assert_eq!(idea.context, None);
    }
}

File: tdz/py/idea.py
"""Idea management for Todozi."""

from typing import Dict, List, Optional
from datetime import datetime
import uuid as uuid_lib

from .models import Idea, IdeaImportance, ShareLevel, ItemStatus
from .error import ValidationError


class IdeaManager:
    """Manages ideas."""

    def __init__(self):
        self.ideas: Dict[str, Idea] = {}
        self.idea_tags: Dict[str, List[str]] = {}

    async def create_idea(self, idea: Idea) -> str:
        """Create a new idea."""
        idea.id = str(uuid_lib.uuid4())
        idea.created_at = datetime.utcnow()
        idea.updated_at = datetime.utcnow()
        self.idea_tags[idea.id] = idea.tags.copy()
        self.ideas[idea.id] = idea
        return idea.id

    def get_idea(self, idea_id: str) -> Optional[Idea]:
        """Get an idea by ID."""
        return self.ideas.get(idea_id)

    def get_all_ideas(self) -> List[Idea]:
        """Get all ideas."""
        return list(self.ideas.values())

    async def update_idea(self, idea_id: str, updates: 'IdeaUpdate'):
        """Update an idea."""
        idea = self.ideas.get(idea_id)
        if not idea:
            raise ValidationError(f"Idea {idea_id} not found")

        if updates.idea is not None:
            idea.idea = updates.idea
        if updates.share is not None:
            idea.share = updates.share
        if updates.importance is not None:
            idea.importance = updates.importance
        if updates.tags is not None:
            idea.tags = updates.tags.copy()
            self.idea_tags[idea_id] = updates.tags
        if updates.context is not None:
            idea.context = updates.context

        idea.updated_at = datetime.utcnow()

    async def delete_idea(self, idea_id: str):
        """Delete an idea."""
        if idea_id not in self.ideas:
            raise ValidationError(f"Idea {idea_id} not found")
        del self.ideas[idea_id]
        self.idea_tags.pop(idea_id, None)

    def search_ideas(self, query: str) -> List[Idea]:
        """Search for ideas."""
        query_lower = query.lower()
        return [
            idea for idea in self.ideas.values()
            if query_lower in idea.idea.lower()
            or any(query_lower in tag.lower() for tag in idea.tags)
            or (idea.context and query_lower in idea.context.lower())
        ]

    def get_ideas_by_importance(self, importance: IdeaImportance) -> List[Idea]:
        """Get ideas by importance level."""
        return [i for i in self.ideas.values() if i.importance == importance]

    def get_ideas_by_share_level(self, share_level: ShareLevel) -> List[Idea]:
        """Get ideas by share level."""
        return [i for i in self.ideas.values() if i.share == share_level]

    def get_ideas_by_tag(self, tag: str) -> List[Idea]:
        """Get ideas by tag."""
        tag_lower = tag.lower()
        return [
            i for i in self.ideas.values()
            if any(t.lower() == tag_lower for t in i.tags)
        ]

    def get_public_ideas(self) -> List[Idea]:
        """Get public ideas."""
        return self.get_ideas_by_share_level(ShareLevel.PUBLIC)

    def get_team_ideas(self) -> List[Idea]:
        """Get team ideas."""
        return self.get_ideas_by_share_level(ShareLevel.TEAM)

    def get_private_ideas(self) -> List[Idea]:
        """Get private ideas."""
        return self.get_ideas_by_share_level(ShareLevel.PRIVATE)

    def get_breakthrough_ideas(self) -> List[Idea]:
        """Get breakthrough ideas."""
        return self.get_ideas_by_importance(IdeaImportance.BREAKTHROUGH)

    def get_recent_ideas(self, limit: int) -> List[Idea]:
        """Get recent ideas."""
        ideas = sorted(self.ideas.values(), key=lambda x: x.created_at, reverse=True)
        return ideas[:limit]

    def get_all_tags(self) -> List[str]:
        """Get all unique tags."""
        all_tags = set()
        for tags in self.idea_tags.values():
            all_tags.update(tags)
        return list(all_tags)

    def get_tag_statistics(self) -> Dict[str, int]:
        """Get tag usage statistics."""
        stats = {}
        for tags in self.idea_tags.values():
            for tag in tags:
                stats[tag] = stats.get(tag, 0) + 1
        return stats

    def get_idea_statistics(self) -> 'IdeaStatistics':
        """Get idea statistics."""
        return IdeaStatistics(
            total_ideas=len(self.ideas),
            public_ideas=len(self.get_public_ideas()),
            team_ideas=len(self.get_team_ideas()),
            private_ideas=len(self.get_private_ideas()),
            breakthrough_ideas=len(self.get_breakthrough_ideas()),
            unique_tags=len(self.get_all_tags())
        )


class IdeaUpdate:
    """Idea update structure."""
    def __init__(self):
        self.idea: Optional[str] = None
        self.share: Optional[ShareLevel] = None
        self.importance: Optional[IdeaImportance] = None
        self.tags: Optional[List[str]] = None
        self.context: Optional[str] = None


class IdeaStatistics:
    """Idea statistics."""
    def __init__(self, total_ideas: int, public_ideas: int, team_ideas: int,
                 private_ideas: int, breakthrough_ideas: int, unique_tags: int):
        self.total_ideas = total_ideas
        self.public_ideas = public_ideas
        self.team_ideas = team_ideas
        self.private_ideas = private_ideas
        self.breakthrough_ideas = breakthrough_ideas
        self.unique_tags = unique_tags

    def public_percentage(self) -> float:
        """Get percentage of public ideas."""
        return (self.public_ideas / self.total_ideas * 100.0) if self.total_ideas > 0 else 0.0

    def team_percentage(self) -> float:
        """Get percentage of team ideas."""
        return (self.team_ideas / self.total_ideas * 100.0) if self.total_ideas > 0 else 0.0

    def private_percentage(self) -> float:
        """Get percentage of private ideas."""
        return (self.private_ideas / self.total_ideas * 100.0) if self.total_ideas > 0 else 0.0

    def breakthrough_percentage(self) -> float:
        """Get percentage of breakthrough ideas."""
        return (self.breakthrough_ideas / self.total_ideas * 100.0) if self.total_ideas > 0 else 0.0


def parse_idea_format(idea_text: str) -> Idea:
    """Parse idea from XML-like format."""
    start_tag = "<idea>"
    end_tag = "</idea>"

    start = idea_text.find(start_tag)
    if start == -1:
        raise ValidationError("Missing <idea> start tag")

    end = idea_text.find(end_tag)
    if end == -1:
        raise ValidationError("Missing </idea> end tag")

    content = idea_text[start + len(start_tag):end]
    parts = [s.strip() for s in content.split(';')]

    if len(parts) < 3:
        raise ValidationError("Invalid idea format: need at least 3 parts (idea; share; importance)")

    # Parse share level
    share_str = parts[1].lower()
    if share_str in ["share", "public"]:
        share = ShareLevel.PUBLIC
    elif share_str in ["dont share", "don't share", "private"]:
        share = ShareLevel.PRIVATE
    elif share_str == "team":
        share = ShareLevel.TEAM
    else:
        share = ShareLevel.PRIVATE

    # Parse tags
    tags = []
    if len(parts) > 3 and parts[3]:
        tags = [s.strip() for s in parts[3].split(',')]

    # Parse context
    context = parts[4] if len(parts) > 4 and parts[4] else None

    return Idea(
        id=str(uuid_lib.uuid4()),
        idea=parts[0],
        project_id=None,
        status=ItemStatus.ACTIVE,
        share=share,
        importance=IdeaImportance(parts[2].lower()),
        tags=tags,
        context=context,
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow()
    )
